test/api.test.js

```diff
@@ -0,0 +1,292 @@
+  });
+
+  test('forbidden database names', async () => {
+    const data = await request.get('/db/_a$aa/entry/aaa').expect(403);
+    expect(data.body).toMatchObject({
+      error: 'invalid database name',
+      code: 'forbidden',
+    });
+  });
+});
+
+describe('rest-api as b@b.com (noRights)', () => {
+  beforeAll(async () => {
+    await noRights();
+    await authenticateAs(request, 'b@b.com', '123');
+  });
+
+  test('query view with owner, wrong key', async () => {
+    const rows = await request
+      .get('/db/test/_query/entryIdByRight?key=xxx')
+      .expect(200);
+    expect(rows.text).toBe('[]');
+  });
+
+  test('query view with owner', async () => {
+    const rows = await request
+      .get(
+        `/db/test/_query/entryIdByRight?key=${encodeURIComponent(
+          JSON.stringify(['x', 'y', 'z']),
+        )}`,
+      )
+      .expect(200);
+    expect(rows.text).not.toBe('[]');
+  });
+
+  test('get an entry authenticated', async () => {
+    await request.get('/db/test/entry/onlyB').expect(200);
+  });
+
+  test('get an entry authenticated but asAnonymous', async () => {
+    await request.get('/db/test/entry/onlyB?asAnonymous=true').expect(401);
+  });
+
+  test('check if user has read access to a resource', async () => {
+    const data = await request
+      .get('/db/test/entry/onlyB/_rights/read')
+      .expect(200);
+    expect(data.body).toBe(true);
+  });
+
+  test('check if user has write access to a resource (as anonymous)', async () => {
+    const data = await request
+      .get('/db/test/entry/onlyB/_rights/read?asAnonymous=1')
+      .expect(200);
+    expect(data.body).toBe(false);
+  });
+});
+
+describe('rest-api as anonymous (data)', () => {
+  beforeAll(async () => {
+    await data();
+    await logout(request);
+  });
+
+  test('save and delete an attachment', async () => {
+    await authenticateAs(request, 'a@a.com', '123');
+    // The attachment name intentionally has an encoded + to make sure URLs are
+    // correctly encoded and decoded by the API.
+    let res = await request
+      .put('/db/test/entry/B/my%2Battachment.txt')
+      .set('Content-Type', 'text/plain')
+      .set('Accept', 'application/json')
+      .send('rest-on-couch!!')
+      .expect(200);
+
+    expect(res.body.id).toBe('B');
+    expect(res.body.rev).toMatch(/^2/);
+    await logout(request);
+
+    const doc = await request.get('/db/test/entry/B');
+    expect(doc.body._attachments['my+attachment.txt']).not.toBe(undefined);
+
+    res = await request.get('/db/test/entry/B/my%2Battachment.txt');
+    expect(res.text).toBe('rest-on-couch!!');
+    expect(res.headers['content-type']).toBe('text/plain');
+
+    await authenticateAs(request, 'a@a.com', '123');
+    await request
+      .delete('/db/test/entry/B/my%2Battachment.txt')
+      .send()
+      .expect(200);
+    await logout(request);
+
+    await request.get('/db/test/entry/B/my%2Battachment.txt').expect(404);
+  });
+});
```

Comment:
I don't see how the test could could have ever worked before without this authentication dance.

Anonymous users don't have permission to create/delete attachments: https://github.com/cheminfo/rest-on-couch/blob/ee0082e587f05e097ddec92ae80cd5406d643005/test/data/data.js

Reply:
I agree with you